### PR TITLE
LLL: end to end testing

### PR DESCRIPTION
**Pull request by zigguratt**
Added tests for `panic` and `exp`.

### CODE REVIEW

**Comment by axic** [comment]
Please use `git rebase`, there should not be any merge commits in pull requests.

**Comment by zigguratt** [comment]
Merge commit removed

File: test/liblll/EndToEndTest.cpp [hunk 1]
```diff
@@ -38,9 +38,93 @@ BOOST_FIXTURE_TEST_SUITE(LLLEndToEndTest, LLLExecutionFramework)
 
 BOOST_AUTO_TEST_CASE(smoke_test)
 {
-        char const* sourceCode = "(returnlll { (return \"test\") })";
-        compileAndRun(sourceCode);
-        BOOST_CHECK(callFallback() == encodeArgs(string("test", 4)));
+	char const* sourceCode = "(returnlll { (return \"test\") })";
```
Comment (axic) [inline test/liblll/EndToEndTest.cpp]: I think you're changing indentation throughout the file?

File: test/liblll/EndToEndTest.cpp [hunk 2]
```diff
@@ -43,6 +43,90 @@ BOOST_AUTO_TEST_CASE(smoke_test)
...
+			(codecopy 0x00 (bytecodesize) 64)
+			(sstore 0x00 @0x00)
+			(sstore 0x01 @0x20)
+			(returnlll
+				(seq
+					(when (= (div (calldataload 0x00) (exp 2 224)) 0xf2c9ecd8)
+						(return @@0x00))
+					(when (= (div (calldataload 0x00) (exp 2 224)) 0x89ea642f)
+						(return @@0x01))
+					(jump 0x02))))
```
Comment (axic) [inline test/liblll/EndToEndTest.cpp]: If there's a "panic" here an appropriate test should be added `callFallback()` and even `callContractFunction("invalidFunction()")`.

Probably though this should split into two tests:
- constructor argument saving (this current one, which doesn't need a `panic`)
- `when` waterfall

**Comment by axic** [comment]
We should also have tests for the arithmetic/binary/unary operations:
```
                std::map<std::string, Instruction> const c_arith = {
                        { "+", Instruction::ADD },
                        { "-", Instruction::SUB },
                        { "*", Instruction::MUL },
                        { "/", Instruction::DIV },
                        { "%", Instruction::MOD },
                        { "&", Instruction::AND },
                        { "|", Instruction::OR },
                        { "^", Instruction::XOR }
                };
                std::map<std::string, pair<Instruction, bool>> const c_binary = {
                        { "<", { Instruction::LT, false } },
                        { "<=", { Instruction::GT, true } },
                        { ">", { Instruction::GT, false } },
                        { ">=", { Instruction::LT, true } },
                        { "S<", { Instruction::SLT, false } },
                        { "S<=", { Instruction::SGT, true } },
                        { "S>", { Instruction::SGT, false } },
                        { "S>=", { Instruction::SLT, true } },
                        { "=", { Instruction::EQ, false } },
                        { "!=", { Instruction::EQ, true } }
                };
                std::map<std::string, Instruction> const c_unary = {
                        { "!", Instruction::ISZERO },
                        { "~", Instruction::NOT }
                };
```

**Comment by pirapira** [comment]
Restarting appveyor.

File: test/liblll/EndToEndTest.cpp [hunk 3]
```diff
@@ -43,6 +43,235 @@ BOOST_AUTO_TEST_CASE(smoke_test)
...
+		fromHex("0100010100010001010001000101000100010100010001000000000000000000")));
+}
+
+BOOST_AUTO_TEST_CASE(unary)
+{
+	char const* sourceCode = R"(
+		(returnlll
+			(seq
+				(mstore8 0x00 (! (< 53 87)))
+				(mstore9 0x01 (! (>= 42 73)))
```
Comment (axic) [inline test/liblll/EndToEndTest.cpp]: `mstore9` :)

**Review by chriseth — APPROVED** [review]

**Pull request merged**